pkg/engine/resolve/dataloader.go

```diff
@@ -0,0 +1,119 @@
+package resolve
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/buger/jsonparser"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/fastbuffer"
+)
+
+type DataLoader struct {
+	totalNum           int
+	inputs             [][]byte
+	inputPosToBufPairs map[int]*BufPair
+	fetch              *BatchFetch
+
+	batch *batch
+
+	mu *sync.Mutex
+}
+
+type batch struct {
+	err  error
+	done chan struct{}
+}
+
+func (d *DataLoader) Load(ctx *Context, input []byte, bufPair *BufPair) (err error) {
+	d.mu.Lock()
+	currentPosition := len(d.inputs)
+	d.inputPosToBufPairs[currentPosition] = bufPair
+	d.inputs = append(d.inputs, input)
+
+	if d.batch == nil {
+		d.batch = &batch{done: make(chan struct{})}
+	}
+
+	d.mu.Unlock()
+
+	if currentPosition == (d.totalNum - 1) { // The last element of the batch
+		go d.resolveFetch(ctx)
+	}
+
+	fmt.Println("before waiting", d.totalNum)
+	<-d.batch.done
```

Comment:
we need to select the context deadline here,
that is, cancel the batch fetch if context was cancelled